[Elixir] New concept exercise: maps (#1336)
exercism/v3: languages/elixir/exercises/concept/maps/.docs/introduction.md

Maps in Elixir are the goto data structure for storing information in key-value pairs. In other languages, these might also be known as associative arrays (PHP), hashes (Perl 5, Raku), or dictionaries (Python).

> Keys and values can be of any data type, but if the key is an atom we can make use of special access protocols. Maps do not guarantee the order of their entries when accessed or returned.

Reviewer: > we can make use of special access protocols

I, as a non-Elixir person, was slightly confused by what this means, especially the protocols bit. Maybe this can be phrased to something like:

```suggestion
Keys and values can be of any data type, but if the key is an atom we can use a shorthand syntax. Maps do not guarantee the order of their entries when accessed or returned.
```

Maps in Elixir are the goto data structure for storing information in key-value pairs. In other languages, these might also be known as associative arrays (PHP), hashes (Perl 5, Raku), or dictionaries (Python).

> Keys and values can be of any data type, but if the key is an atom we can make use of special access protocols. Maps do not guarantee the order of their entries when access or returned.

Reviewer: ```suggestion
Keys and values can be of any data type, but if the key is an atom we can make use of special access protocols. Maps do not guarantee the order of their entries when accessed or returned.
```

> > Elixir gotcha! When an instance of a map has a small number of entries they appear to be ordered because of their internal implementation as an ordered keyword list, when their size exceeds 32 items, it is then stored as a [hash array mapped tree][hamt].

Reviewer: I would suggest to remove this gotcha here and just have it in the `after.md` document, as I think the student does not need to know this to be able to complete the exercise, right?

Author: Agreed, I think I wrote this and intended to remove it.  Then didn't! Thanks for catching it.

```suggestion
```

## Literal forms

An empty map is simply declared with `%{}`. If we want to add items to a map literal, we can use two forms:

```elixir
# If the key is an atom:
%{atom_key: 1}

# If the key is a different type:
%{1 => :atom_value}

# You can even mix these if the atom form comes second:
%{"first_form" => :a, atom_form: :b}
```

> It would be convention to choose a consistent form which first all of the data in the literal.

Reviewer: I'm having a slightly hard time understanding this sentence. Is there possibly a word missing?

Author: ```suggestion
While there is no canonical format, choose a consistent way to represent the key-value literal pairs.
```

Author: Does that make more sense?

Reviewer: Yes, I like that!

## Map module functions

> Elixir provides many functions for working with maps to access/get/put/remove/drop values and keys.

Reviewer: Instead of using several examples of `Map` module functions, I would suggest to rephrase to mention that the `Map` module has many functions to work with maps.

Author: ```suggestion
Elixir provides many functions for working with maps in the _Map module_.
```

Reviewer: I like that!

> ```shell

Reviewer: I don't think any new syntax is introduced here, right? If correct, I'd suggest to remove the example entirely.

iex> map = %{}
%{}
iex> map_a = Map.put(map, :a, 1)
%{a: 1}
iex> Map.get(map_a, :a)
1
```

> Some functions require an anonymous or captured function to be passed into the function:

Reviewer: As one of the prerequisites is `anonymous-functions`, we can expect students to already know about anonymous functions, right? If so, I think this bit of text can be removed.

> ```shell

Reviewer: I don't think any new syntax is introduced here, right? If correct, I'd suggest to remove the example entirely.

iex> increment = fn value -> value + 1 end
#Function<7.126501267/1 in :erl_eval.expr/5>
iex> empty_map = %{}
%{}
iex> a_map = Map.update(empty_map, :a, 0, increment)
%{a: 0}
iex> Map.update(a_map, :a, 0, increment)
%{a: 1}
```

## Maps as an `Enumerable` collection

Maps implement the _Enumerable_ protocol, which allow them to be transformed into a list or key-value tuples. Often maps are transformed into a correlating list, processed or transformed in some way, then returned.

> ```shell

Reviewer: I don't think any new syntax is introduced here, right? If correct, I'd suggest to remove the example entirely and perhaps add a mention that the `Enum` module has functions to work with types implementing the _Enumerable_ protocol.

iex> map = %{a: 1, b: 2, c: 3}
%{a: 1, b: 2, c: 3}
iex> list = Map.to_list(map)
[a: 1, b: 2, c: 3]
iex> hd(list)
{:a, 1}
iex> Enum.map(list, fn pair -> (pair |> elem(1)) + 1 end)
[2, 3, 4]
```

## Module attributes as constants

In elixir, we can define module attributes which can be used as constants in our functions.

```elixir
defmodule Example do

  # Defines the attribute as the value 1

>   @constant_value 1

Reviewer: Maybe to be consistent with `@constant_list`, name this:

```suggestion
  @constant_number 1
```

?

>   # Overwrites the attribute with the value 2

Reviewer: Is the ability to overwrite an attribute necessary for the student to know? If not, I would suggest to only mention this in the `after.md` file.

  @constant_value 2

  @constant_list [1, 2, 3]

  def example_value() do
    # Returns the value 2
    @constant_value
  end

>   def example_list() do

Reviewer: Maybe remove this example as the above example already shows how to use constants.

    # Returns the list [1, 2, 3]
    @constant_list
  end
end
```

> When used this way, attributes can be any expression which can be evaluated at compilation time. After compilation, module attributes are not accessible since they are exanded during cimpilation, similar to defined macros in languages like C.

Reviewer: The "When used this way" bit seems to imply that there is another way to use/define attributes. Is this true? You could consider removing that bit entirely.

Author: ```suggestion
When used to define module constants, attributes can be any expression which can be evaluated at compilation time. After compilation, module attributes are not accessible since they are exanded during cimpilation, similar to defined macros in languages like C.
```

Yes there are other uses for module attributes, is this less leading?

Reviewer: Yes! 👍

> When used this way, attributes can be any expression which can be evaluated at compilation time. After compilation, module attributes are not accessible since they are exanded during cimpilation, similar to defined macros in languages like C.

Reviewer: ```suggestion
When used this way, attributes can be any expression which can be evaluated at compilation time. After compilation, module attributes are not accessible since they are expanded during compilation, similar to defined macros in languages like C.
```

Author: see above
